### ISSUE TITLE
Checkpoint loading not working correctly

### ISSUE THREAD

**Issue by sdtblck** [microsoft/DeepSpeed#982]
We are having a critical issue in [gpt-neox](https://github.com/EleutherAI/gpt-neox/) where the checkpoint loading from deepspeed isn't working correctly. When we reload the model from a checkpoint, the loss suggests it hasn't loaded anything at all, and is just starting again from initialization... (jumps up from ~4 at checkpoint save to 8 at checkpoint load)

The model is based on the megatron example in DeepSpeedExamples, and the checkpoint saving / loading as far as I'm aware, hasn't changed at all from there (and should be handled almost entirely by model_engine.save_checkpoint / load_checkpoint.)

It happens both with the pipeline parallel module, and without.

You can see here an example of training loss before saving a checkpoint

![Screenshot from 2021-04-20 13-28-27](https://user-images.githubusercontent.com/46172032/115389106-f9c13d00-a1dc-11eb-95cc-783e459a6d25.png)

and then the loss directly afterward:

![Screenshot from 2021-04-20 13-30-55](https://user-images.githubusercontent.com/46172032/115389130-00e84b00-a1dd-11eb-9516-0c6111490b40.png)

Would appreciate any help to decipher what's going on here.

We're technically using our [deepspeed fork](https://github.com/EleutherAI/DeeperSpeed) but there are no modification to the checkpoint saving / loading logic there either.

This is about as minimal a reproduction as i can give right now:

first set up the environment in bash:
(this is assuming you're running on an aws machine, but should work on any base image with torch + cuda 11.1, just comment out that first line)

```bash
# first activate the correct AWS env with torch 1.8.0 + cuda 11.1
source activate pytorch_latest_p37

# clone repo and cd into it
git clone https://github.com/EleutherAI/gpt-neox
cd gpt-neox

# install other requirements
pip install pybind11==2.6.2 six regex numpy==1.20.1 nltk==3.5 \
        -e git+git://github.com/EleutherAI/DeeperSpeed.git@0e9573776caf7227ecafdc8a3d57c7955165d8e2#egg=deepspeed \
        zstandard==0.15.1 cupy-cuda111 mpi4py==3.0.3 wandb==0.10.21 einops==0.3.0 transformers tokenizers lm_dataformat triton==1.0.0.dev20210329

# install apex
pip install -v --disable-pip-version-check --no-cache-dir --global-option="--cpp_ext" --global-option="--cuda_ext" git+https://github.com/NVIDIA/apex.git@e2083df5eb96643c61613b9df48dd4eea6b07690

# prepare dummy gpt2 data
python prepare_data.py
```

then in a python script, train a small model and try to continue training:
```python
from yaml import load, dump
try:
    from yaml import CLoader as Loader, CDumper as Dumper
except ImportError:
    from yaml import Loader, Dumper
import os

# modify train iters in small.yml to 500
with open('configs/small.yml', 'r') as f:
        data = load(f, Loader=Loader)
data.update({'train-iters': 500})
with open('configs/tmp.yml', 'w') as f:
        f.write(dump(data))

# start run
os.system('./deepy.py pretrain_gpt2.py -d configs tmp.yml local_setup.yml')

# kill any hanging processes
os.system("pkill -f deepy.py")

# update train iters again
with open('configs/small.yml', 'r') as f:
        data = load(f, Loader=Loader)
data.update({'train-iters': 550})
with open('configs/tmp.yml', 'w') as f:
        f.write(dump(data))

# start run again
os.system('./deepy.py pretrain_gpt2.py -d configs tmp.yml local_setup.yml')
```

**Comment by sdtblck** [comment]
so I decided to check that the weights had actually changed after loading from the checkpoint like so:
```
        before_load = model.module.language_model.transformer.layers[0].attention.dense._parameters["weight"]
        checkpoint_name, state_dict = model.load_checkpoint(load_dir)
        after_load = model.module.language_model.transformer.layers[0].attention.dense._parameters["weight"]
        eq = (before_load == after_load).all()
        print(eq)
```

and it returns `tensor(True, device='cuda:2')` for all devices. This suggests to me that checkpoint loading hasn't actually done anything - but it also doesn't produce any error.

**Comment by sdtblck** [comment]
Just verified that the weights in the checkpoint aren't the same as those in the model by entering pdb after `        checkpoint = torch.load(load_path, map_location=lambda storage, loc: storage)` in deepspeed engine and evaluating: 

`p (checkpoint["module"]["language_model"]["transformer"]["layers.0.attention.dense.weight"].cpu() == self.module.language_model.transformer.layers[0].attention.dense._parameters["weight"].cpu()).all()` 

output:
`tensor(False)`

:thinking:

**Comment by sdtblck** [comment]
> so I decided to check that the weights had actually changed after loading from the checkpoint like so:
> 
> ```python
> before_load = model.module.language_model.transformer.layers[0].attention.dense._parameters["weight"]
> checkpoint_name, state_dict = model.load_checkpoint(load_dir)
> after_load = model.module.language_model.transformer.layers[0].attention.dense._parameters["weight"]
> eq = (before_load == after_load).all()
> print(eq)
> ```
> 
> and it returns `tensor(True, device='cuda:<deviceno>')` for all devices. This suggests to me that checkpoint loading hasn't actually done anything - but it also doesn't produce any error.

Doing this same thing before and after self.load_module_state_dict gives the same output, so I would guess the error is somewhere in self.module.load_state_dict? very confusing

**Comment by sdtblck** [comment]
Okay so I'm being dumb and forgot to `.clone()` the tensors - but there is still some problem.

Each new run on this graph is reloading the checkpoints and restarting training - so you can see it does *sort of* work, but there seems to be some problem with the very first reload?

![Screenshot from 2021-04-20 17-14-43](https://user-images.githubusercontent.com/46172032/115421174-07d28600-a1fc-11eb-93d7-ae0f00e53adf.png)

**Comment by sdtblck** [comment]
I can verify that this also happens with the Megatron3D-Parallelism example in DeepSpeedExamples, without any of our modifications.

![Screenshot from 2021-04-21 18-51-29](https://user-images.githubusercontent.com/46172032/115591719-b5b06400-a2d2-11eb-9bc7-409dd485beac.png)

the above is the loss of the `ds_pretrain_gpt2.sh` script in https://github.com/microsoft/DeepSpeedExamples/tree/master/Megatron-LM-v1.1.5-3D_parallelism after saving / loading

**Comment by tjruwase** [comment]
@sdtblck. sorry about this issues. Will take a look. Is your small repro a great place to start?

**Comment by sdtblck** [comment]
Hi @tjruwase I think so, let me know if you have any problems setting it up. As you can see I also reproduced it with the megatron examples, so if that works better for you then that would work too. 

Only problem is i can't reproduce it consistently. Sometimes it works as expected, and sometimes it doesn't. 

We have a big run for which we can't load the checkpoints at all, they're basically reinitializing.

**Comment by sdtblck** [comment]
I have a feeling the larger run may be to do with https://github.com/microsoft/DeepSpeed/issues/787 but i'm also struggling to reproduce that error consistently

**Comment by sdtblck** [comment]
Hi @tjruwase - I found the bug. 

Actually nothing to do with deepspeed, but it does also affect the megatron examples in deepspeed examples, so you might want to fix it. 

Fix described here -> https://github.com/EleutherAI/gpt-neox/pull/251

**Comment by tjruwase** [comment]
@sdtblck, thanks for hard work that went into resolving this.  Can you please close this issue?
